### PR TITLE
Show actual linked color in color buttons when color has been linked to a project color

### PR DESCRIPTION
**Pull request by nyalldawson**
Instead of only showing linked color status in the accompanying
data defined button, we now also show it inside the color button
itself. So now the button color is a live reflection of the
project color it is linked to.

Additionally, when the color is linked, the button only allows
users the choice of "unlinking" the color from the linked
project color. The usual color button options allowing users
to directly change the color are not present. (As linked
colors must be edited in the project colors section from
the project properties dialog)

### CODE REVIEW

**Comment by nyalldawson** [comment]
@nirvn, for your review please :smile:

File: src/gui/qgscolorbutton.cpp [hunk 1]
```diff
@@ -443,14 +495,21 @@ QPixmap QgsColorButton::createMenuIcon( const QColor &color, const bool showChec
 
 void QgsColorButton::buttonClicked()
 {
-  switch ( mBehavior )
+  if ( !mLinkedColorName.isEmpty() )
```
Comment (nirvn) [inline src/gui/qgscolorbutton.cpp]: For the sake of consistency, we should probably settled on either mLinkedColorName.isEmpty() or  linkedProjectColor().isValid() as a conditional check for a linked project color. Whichever is less costly :) see e.g. QgsColorButton::dragEnterEvent for the alternate check.

File: src/gui/qgscolorbutton.cpp [hunk 2]
```diff
@@ -462,95 +521,110 @@ void QgsColorButton::prepareMenu()
...
+
+    if ( mColorSchemeRegistry )
     {
-      QgsColorSwatchGridAction *colorAction = new QgsColorSwatchGridAction( *it, mMenu, mContext, this );
-      colorAction->setBaseColor( mColor );
-      mMenu->addAction( colorAction );
-      connect( colorAction, &QgsColorSwatchGridAction::colorChanged, this, &QgsColorButton::setValidColor );
-      connect( colorAction, &QgsColorSwatchGridAction::colorChanged, this, &QgsColorButton::addRecentColor );
+      //get schemes with ShowInColorButtonMenu flag set
+      QList< QgsColorScheme * > schemeList = mColorSchemeRegistry->schemes( QgsColorScheme::ShowInColorButtonMenu );
```
Comment (nirvn) [inline src/gui/qgscolorbutton.cpp]: Should there be a special case for project color swatch that'd match the behavior of that of the data-defined property button, or do we not lock to the project color when selecting it via this swatch?

File: src/gui/qgscolorbutton.cpp [hunk 3]
```diff
@@ -615,8 +689,18 @@ void QgsColorButton::addRecentColor( const QColor &color )
...
 {
   QColor backgroundColor = color;
-
-  if ( !color.isValid() )
+  bool isProjectColor = false;
+  if ( !backgroundColor.isValid() && !mLinkedColorName.isEmpty() )
+  {
+    backgroundColor = linkedProjectColor();
+    isProjectColor = backgroundColor.isValid();
+    if ( !isProjectColor )
```
Comment (nirvn) [inline src/gui/qgscolorbutton.cpp]: Do you need to rely on this bool, or could you just use backgroundColor.isValid() here?

File: src/gui/qgscolorbutton.cpp (hunk 2 above)
Comment (nirvn) [reply src/gui/qgscolorbutton.cpp]: Here's an implementation idea: add a linked( QString &name ) signal, and connect this signal to the data defined property widget in the registerLinkedWidget(...) function.

**Review by nirvn — COMMENTED** [review]
Nice, nice!

File: src/gui/qgscolorbutton.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/gui/qgscolorbutton.cpp]: We don't lock when selecting from swatches -- it just sets to that color. There's no way to create a linked color apart from via data defined properties/the data defined button.

File: src/gui/qgscolorbutton.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/gui/qgscolorbutton.cpp]: I was going to use it later, when/if we decide on drawing an icon overlay for linked colors

File: src/gui/qgscolorbutton.cpp (hunk 2 above)
Comment (nirvn) [reply src/gui/qgscolorbutton.cpp]: @nyalldawson , it's fine by me if we decide not to lock onto a project color if we select it via the color button swatch, I can see how that logic can be useful.

**Review by nirvn — APPROVED** [review]

**Pull request merged**